ENH: optimize: `least_squares` LM Method with Automatic Jacobian Scaling when scaling factor is not specified (Default `diag=None`).

Roshan-Velpula opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message

Depending on your changes, you can skip CI operations and save time and energy: 
http://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
This PR addresses the issue described in #19459 - "BUG: optimize.least_squares giving poor result compared to optimize.leastsq and optimize.curve_fit".

#### What does this implement/fix?
This enhancement modifies the behavior of optimize.least_squares when used with the method='lm' (Levenberg-Marquardt algorithm). Previously, unless x_scale was set to 'jac', the default behavior was to use diag=1.0 for scaling, leading to suboptimal fits in certain scenarios. This change proposes to set the default value of x_scale='jac' when method=='lm', aligning the behavior more closely with that of optimize.leastsq and optimize.curve_fit. This results in improved optimization performance and robustness.
Summary of Changes
Set default x_scale='jac' for least_squares with method='lm'.

#### Additional information
<!--Any additional information you think is important.-->

Roshan-DSBA committed:
changes to least_squares

j-bowhay commented:
Can you add a unit test demonstrating that the change performs as intended?

Roshan-DSBA committed:
Improve LM algorithm scaling in least_squares

This commit enhances the behavior of the Levenberg-Marquardt (LM) algorithm in the least_squares function. The key change is setting the default value of x_scale to 'jac' when the LM method is used. This modification aligns the LM method more closely with the behaviors of other optimization algorithms in SciPy and addresses issues related to poor optimization performance when dealing with parameters of varying scales. The update aims to enhance the robustness and efficiency of the LM method, particularly in scenarios with significant scale variation among parameters.

Add unit tests for LM method with default and no scaling

This commit introduces unit tests for the LM method in the least_squares function. The tests are designed to assess the behavior of the method with the default scaling (x_scale set to 'jac') and without scaling (x_scale explicitly set to 1.0). The addition of these tests is crucial for evaluating the effectiveness of the recent fix in the LM algorithm, which sets the default scaling to 'jac' to improve robustness and accuracy. The tests compare the performance with and without this default scaling after the fix proposed, ensuring that the algorithm behaves as expected and the fix addresses the identified issue.

Roshan-Velpula commented:
> Can you add a unit test demonstrating that the change performs as intended?

Updated the unit test and committed the change. Please review the PR and let me know.

Roshan-DSBA committed:
In a follow-up to my previous commit, this update addresses an oversight regarding the default behavior of the least_squares function when using the Levenberg-Marquardt (LM) method. In cases where x_scale was unspecified, the function did not default to Jacobian scaling (diag = None) as intended. This commit corrects that behavior, ensuring that the LM method now correctly defaults to Jacobian scaling when x_scale is None.

dschmitz89 commented on scipy/optimize/_lsq/least_squares.py:
```diff
@@ -47,10 +47,15 @@ def call_minpack(fun, x0, jac, ftol, xtol, gtol, max_nfev, x_scale, diff_step):
     # Compute MINPACK's `diag`, which is inverse of our `x_scale` and
     # ``x_scale='jac'`` corresponds to ``diag=None``.
-    if isinstance(x_scale, str) and x_scale == 'jac':
+    if method == 'lm' and x_scale == None:
```
```suggestion
    if method == 'lm' and x_scale is None:
```

dschmitz89 reviewed (changes requested):
test

Roshan-DSBA committed:
updated x_scale = None or x_scale is None
Removed test_default_scaling and test_no_scaling test cases as the test_comparison test case demonstrates the improvement

Roshan-Velpula commented:
> I cannot find how to remove my review here unfortunately. Will take a closer look in some time.

Thanks for the email and suggested changes, I did them accordingly and committed them. Please let me know if you need any further updates from me :)

lucascolley commented:
Hi @Roshan-DSBA , if you do the following the doc build check should work:

```
git fetch --all
git merge upstream/main
git push
```

Roshan-DSBA committed:
Merge remote-tracking branch 'upstream/main'

Roshan-Velpula commented:
> Hi @Roshan-Velpula , if you do the following the doc build check should work:
> 
> ```
> git fetch --all
> git merge upstream/main
> git push
> ```

Hi, Thank you for the message! Pushed the changes and doc build check worked. :)

dschmitz89 reviewed (commented):
The change itself looks good to me but I do not understand how the test demonstrates its correctness. Maybe I am misunderstanding but it looks like `x_scale=1` does not change the Jacobian at all. What about choosing `xscale >>  1`?

CarrascoCesar commented:
Hello @lucascolley is there any interest in updating this PR?  I am interested in revisiting it since it would solve an important issue in my view.

lucascolley commented:
Hey @CarrascoCesar ! This isn't my area of expertise, but perhaps you could respond to @dschmitz89's comment above about the purpose of the test?

CarrascoCesar commented:
Hello @dschmitz89 would there be interest in revisiting the PR and creating a simpler test that checks for improved behavior when x_scale is not defined and least_squares defaults to parameter scaling?

dschmitz89 commented:
@CarrascoCesar : sorry, looking again I think I misunserstood your test. My new interpretation is: you are comparing the new against the old default and the new default has to result in better convergence indicated by a smaller number of function evaluations. Is that correct? In that sense, the test is actually very helpful.

Reading again the complete PR description, it seems that the overall goal is to get the same behaviour as `leastsq` if `least_squares` uses Levenberg Marquardt. Currently this is not the case due to different default arguments. WIth this change, we should get the same results. How about we update the test so that we show that we get the same results for `least_squares(method=lm)` and `leastsq`? This would make the purpose more explicit.

CarrascoCesar commented:
Hello @dschmitz89, I agree, comparing least_square to leastsq in the test addresses the original issue more directly. I'll work on it to include your recommendation and see if I can come up with a simpler test problem.

CarrascoCesar commented:
Hello @lucascolley, this will be my first time modifying somebody else's PR.  What is the process to checkout the commit for me to make changes and then post?

lucascolley commented:
> Hello @lucascolley, this will be my first time modifying somebody else's PR. What is the process to checkout the commit for me to make changes and then post?

hey @CarrascoCesar , personally, I would use GitHub's CLI tool `gh` and do the following:

1. Make sure you are in the directory of your clone of this repo
2. `gh checkout 19700` (to checkout this PR)
3. `git checkout -b LM-changes` (to make a new branch matching this PR)
4. Make your changes and commit as usual
5. Open a new PR from your branch

If you don't want to use `gh`, you can do the same by adding @Roshan-Velpula's fork as a remote.

If @Roshan-Velpula were active and collaborating, you could instead make a PR to their fork, and the changes could stay in this PR. But no harm in just opening a new PR here!

andyfaff commented:
Closed in favour of #22790
